Register the Recommendations block using Block MetaData and a block.json file

ryanwelcher opened the pull request:
## Description
This PR is based on #757 and contains additional changes. It changes and simplifies the block registration process by leveraging block metadata and a `block.json` file, and it solves the following issues:
- Styling issues in FSE
- Block showing error or disappearing in FSE after save and reload

## Motivation and Context
- This change stemmed from issues with this block in FSE and after reviewing the code in an attempt to fix the issue, I made recommendations to change both how the block is registered (reflected in this PR) as well as some changes to the experience in the block editor.
- Fixes #711

## How Has This Been Tested?
Tested locally only

acicovic committed:
Fix styling issues in FSE mode

acicovic committed:
Recommendations Block: Enable in FSE mode

acicovic committed:
Merge branch 'develop' into fix/recommendations-block-fse-issues

acicovic committed:
Recommendations Block: Enable by default

acicovic committed:
Revert "Recommendations Block: Enable by default"

This reverts commit 1014e4250d1176cff3ab02bf32ece5430fd64e7d.

ryanwelcher committed:
Introduce block.json to be used as the source of truth for block definition.

ryanwelcher committed:
Registers the block by passing the path the block.json file.
Removes the supports and attributes items from PHP as they are now manged via block.json.

ryanwelcher committed:
Adjust webpack config to use defaultConfig.entry to take advantage to the automatic entry point detenction and file copying provided by the @wordpress/scripts package.

ryanwelcher committed:
Rename index.js to view.js to better reflect it's usage on the front end only.

ryanwelcher committed:
Add new proprties to the block.json file:
editorScript - defines which file will be enqueue into the block editor. This file handles the block definition on the client side.
style - defines a stylesheet to be enqueue into both the editor and on the front end
viewScript - defines a file that will be registered with the intent of being enqueue on the front end only.

ryanwelcher committed:
Remove the manually registered scripts and styles in favor of the block.json managing them.
Manually enqueue the viewScript defined in block.json.

ryanwelcher committed:
Add the build files for the block.

ryanwelcher committed:
Add textdomain property to block.json.

ryanwelcher committed:
Add admin only CSS and enqueue the file.

ryanwelcher committed:
Add check for 5.9 and fallback viewScript registration if needed.

GaryJones commented on src/blocks/recommendations/class-recommendations-block.php:
```diff
@@ -49,41 +49,38 @@ public function run(): void {
+		if ( ! isset( $wp_version ) || version_compare( $wp_version, '5.9' ) < 0 ) {
+			$plugin_url      = plugin_dir_url( PARSELY_FILE );
+			$view_asset_path = plugin_dir_path( PARSELY_FILE ) . 'build/blocks/recommendations/view.asset.php';
+			if ( file_exists( $view_asset_path ) ) {
```
Assuming that the build should always succeed (and we want a hard fail if it doesn't / hasn't been run locally), what does this condition here give us?

ryanwelcher replied:
It's an extra safeguard against a fatal error if we try to `require` a file that doesn't exist, but that's about it. Happy to remove this check if you feel it's superfluous.

GaryJones replied:
Under what conditions could we reasonably expect the file not to exist, and NOT want a fatal?

ryanwelcher replied:
This would only become an issue when/if the build directory was either non-existent or this file was missing. 

From a developer's point of view, it's not really a big deal to white screen the site if the build didn't get run, but in the (edge)-case of a plugin update going sideways on a client site, white screening the admin would be problematic.  At any rate, I have removed the check.

GaryJones commented on src/blocks/recommendations/block.json:
```diff
@@ -0,0 +1,44 @@
+		},
+		"title": {
+			"type":"string",
+			"default": "Related Content"
```
Can this string be translated at some point?

ryanwelcher replied:
Strings in `block.json` should translated automatically during the block registration process - https://github.com/WordPress/wordpress-develop/blob/573581f43f63f12cc37620172432cf93745f568c/src/wp-includes/blocks.php#L289. I didn't test this however.

GaryJones replied:
If I'm reading that right - that's only for top-level keys, and not an attribute's [default] value.

ryanwelcher replied:
Yes, I think you're correct. Update incoming.

GaryJones commented on src/blocks/recommendations/class-recommendations-block.php:
```diff
@@ -49,41 +49,38 @@ public function run(): void {
-			$script_asset_file['version']
+		register_block_type(
+			plugin_dir_path( PARSELY_FILE ) . 'build/blocks/recommendations/',
+			self::get_block_registration_args()
```
Let's inline the array returned from the `get_blocks_registration_args()` method, and then delete that method.

ryanwelcher committed:
Updates per code review.

GaryJones commented on src/blocks/recommendations/block.json:
```diff
@@ -0,0 +1,44 @@
+	"$schema": "https://schemas.wp.org/wp/5.9/block.json",
+	"apiVersion": 2,
+	"name": "wp-parsely/recommendations",
+	"title": "Parse.ly Recommendations",
```
As per https://core.trac.wordpress.org/changeset/49981 the `title`, `description`, each `keyword`, and each style `label` are automatically wrapped in `translate_with_gettext_context()` with the textdomain passed from the value on line 7 ✅

ryanwelcher committed:
Run build command.

GaryJones commented on src/blocks/recommendations/block.json:
```diff
@@ -0,0 +1,44 @@
+	"editorScript": "file:./edit.js",
+	"style": "file:./style-edit.css",
+	"editorStyle": "file:./edit.css",
+	"viewScript": "file:./view.js"
```
Noting that using / relying on the block.json means WP 5.8 is the minimum, and `viewScript` was only supported in WP 5.9.

jblz replied:
I was trying to remember why I didn't use the json for this in the first place and the minimum WordPress version is it.

ryanwelcher replied:
What is the minimum version this block needs to support?  I probably should have asked that first 😄

acicovic replied:
OK, so the Block in its initial state was supporting WP 5.6 (released December 2020). [That line is still present in class-recommendations-block.php](https://github.com/Parsely/wp-parsely/blob/ed18932ef133a4dd3d8f28492084e43d44e90a78/src/blocks/recommendations/class-recommendations-block.php#L19). Personally, I don’t see a need to support such an old version if this results in difficult problems for us.

The big change with this PR is that _the block now works in FSE_, which wasn’t the case before. I clearly see the point that supporting 5.9 and upwards is only supporting the “bleeding edge”. My suggestion is that if we can support 5.8 with very minor changes that will be easy to update in a few months, we can go that route. _This provided of course that the block will keep working in FSE_.

Else I’m totally fine with supporting 5.9 and upwards, as users who care for their WP installations will sooner or later get there. WordPress 6.0 release is in 24 May and we generally encourage people to upgrade, so 5.9 will not be that new in a couple of months.

acicovic replied:
Regarding the `ViewScript` issue, it seems to me that this should work for 5.8 now with [this commit](https://github.com/Parsely/wp-parsely/pull/762/commits/982a47cf6c785a88af2a91ca01fe217efc588362). WP 5.8 was released on 20 July 2021.

WP 5.9 and 5.8 account for about of 62% of WP installs, with 5.8 being around 11% ([source](https://wordpress.org/about/stats/)). 5.7 and 5.6 follow with a 6% and 3% respectively, so not a huge gain there.

pauarge replied:
We have released the block on version 3.2.0 of the plugin under a feature flag, which gives us the flexibility of doing changes as such. I think Alex is making a very good point about the potential users of the block. 

I am going to argue that the full benefits of the block come with 5.9, so I don't want to commit now to support 5.8, which will probably bring us difficulties down the line. If we all agree, I would strongly push for supporting the block just on 5.9+ and having the least amount of "hacks".

GaryJones replied:
To me, this should be partly driven as a business decision, so much as a "what's easiest for us to maintain" decision.

If we make the whole plugin require WP 5.9, we're potentially cutting off current and potential customers from using it. Folks aren't going to do a whole WP core upgrade just to use a block from wp-parsely - the plugin doesn't have that power.

If we make just the block part require WP 5.9, while the rest of the plugin supports down to 5.0, then it's doable, but it needs clear documentation to set correct expectations.

acicovic replied:
I do agree with the idea that there should be a "decision factor" besides development effort (though effort is definitively a big thing for us as a small team). In cases like this one, I think it should be based on data and goals to help us make correct decisions.

For example, with some statistics on the WP versions used by our users, and (let's say) an intention to keep compatibility with at least 70% of them, this could orient us. Without such rules and data, making correct decisions is harder. Trying to cover almost every user is always a great idea, but it can be costly in practice and limited in returns.

If backwards compatibility becomes problematic for us or for users that do update, I would go for serving the users who update. If we abort the idea of `block.json`, we'll have to find a solution for users who want to have FSE available. But keeping 5.8 compatibility is simple.

All this to say that there is a point in all arguments made here and that we could take this discussion further if needed.

ryanwelcher committed:
Ensuring the defaul value of title is translated.

acicovic committed:
Recommendations Block: Rebuild assets

acicovic commented:
I can confirm from my manual testing that this works fine in WP 5.9.3 (pages, posts, FSE). If we agree to stick with the WP 5.8 support and upwards, we can open this for review officially. Automated testing (as well as other block adjustments) will be addressed in different PRs once this gets merged.

pauarge committed:
Merge branch 'develop' into try/use-block-json-for-block-registration

pauarge commented:
@ryanwelcher The PR looks to me that's on the right path. Do you reckon we could move it out from draft?

acicovic commented:
> @ryanwelcher The PR looks to me that's on the right path. Do you reckon we could move it out from draft?

@pauarge Based on what you said, I think we can remove 5.8 compatibility code and move out of draft (it is I who had set the PR as draft).

acicovic committed:
Merge branch 'develop' into try/use-block-json-for-block-registration

acicovic committed:
Recommendations Block: Update compatibility to WP 5.9 and up

acicovic committed:
Recommendations Block: Remove feature flag

acicovic committed:
README.md: Update Recommendations Block information

acicovic committed:
Recommendations Block README.md: Update Block information

acicovic committed:
Recommendations Block: Simplify editor CSS

acicovic committed:
Recommendations Block: Disable links in editor

acicovic committed:
Add missing period in a comment

acicovic committed:
Merge branch 'develop' into update/remove-recommendations-block-feature-flag

acicovic committed:
Merge branch 'develop' into try/use-block-json-for-block-registration

acicovic committed:
Add a CSS comment for clarity

pauarge committed:
Merge branch 'develop' into update/remove-recommendations-block-feature-flag

pauarge committed:
Merge branch 'develop' into try/use-block-json-for-block-registration

GaryJones commented:
Seems there's a failing integration test for PHP 7.1 that needs reviewing.

pauarge committed:
Merge branch 'develop' into update/remove-recommendations-block-feature-flag

pauarge committed:
Merge branch 'develop' into try/use-block-json-for-block-registration

pauarge commented:
> @GaryJones Seems there's a failing integration test for PHP 7.1 that needs reviewing.

Worth noting that PHP 7.1 is using WordPress 5.8, which might be causing the issue.

acicovic committed:
Recommendations Block: Change URLs to # when in editor

Mouse events are already disabled by CSS, but this was done for better screen reader experience.

acicovic committed:
Merge branch 'develop' into update/remove-recommendations-block-feature-flag

acicovic committed:
Merge branch 'develop' into try/use-block-json-for-block-registration

acicovic committed:
Add filter for enabling/disabling Related API Proxy

acicovic committed:
Roll back to @wordpress/scripts 22.3.0

acicovic committed:
Merge pull request #764 from Parsely/update/remove-recommendations-block-feature-flag

Recommendations Block: Remove feature flag

acicovic committed:
Merge pull request #767 from Parsely/update/recommedations-block-disable-links-in-editor

Recommedations Block: Disable link clicking in Editor

acicovic committed:
Recommendations Block: Improve error handling

pauarge commented on src/blocks/recommendations/components/parsely-recommendations.js:
```diff
@@ -25,11 +25,23 @@ export default function ParselyRecommendations( {
 	} = useRecommendationsStore();
 
+	function getErrorMessage() {
+		errorMessage = __( 'Error: ', 'wp-parsely' ) + JSON.stringify( error );
```
Could we use [template literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) for this? 

Also, `errorMessage` should be declared as a variable.

acicovic committed:
Recommendations Block: Improve code quality

acicovic committed:
Integration Tests: Don't run Block tests against unsupported WP versions

acicovic committed:
Merge branch 'develop' into try/use-block-json-for-block-registration

pauarge commented on README.md:
```diff
@@ -87,7 +87,9 @@ A sample `JSON-LD` meta tag and structured data for an article or post:
 ## The Parse.ly Recommendations Block (beta)
 
-The plugin includes a Recommendations Block that uses the [Parse.ly API](https://www.parse.ly/help/api/recommendations#get-related) to showcase links to content on your site. More information about enabling the Recommendations Block can be found in its [documentation](https://github.com/Parsely/wp-parsely/blob/trunk/src/blocks/recommendations/README.md). This feature is currently in beta and disa ...
+**IMPORTANT:** The Recommendations Block is available only in WordPress 5.9 and later.
```
```suggestion
The Recommendations Block is available only in WordPress 5.9 and later.
```

acicovic replied:
Resolving as agreed.

pauarge commented on src/blocks/recommendations/README.md:
```diff
@@ -1,18 +1,10 @@
-You can add it in `Posts`, `Pages`, or nearly any other custom post type. From WordPress 5.8 and up, the block can also be used as a [Block-based Widget](https://wordpress.org/support/article/block-based-widgets-editor/). Note however, that the Recommendations Block is not available in the Full Site Editor's template editor. This restriction will be lifted in the future.
-
-## Disabled by Default
+**IMPORTANT:** The Recommendations Block is available only in WordPress 5.9 and later.
```
```suggestion
The Recommendations Block is available only in WordPress 5.9 and later.
```

acicovic replied:
Resolving as agreed.

pauarge commented on src/blocks/recommendations/class-recommendations-block.php:
```diff
@@ -49,88 +44,19 @@ public function run(): void {
 	 *
 	 * @return void
```
We should update the `@since` attribute because the function has been renamed.

acicovic replied:
See https://github.com/Parsely/wp-parsely/pull/762/commits/985fdd2cc6604938216cb985a979b5362f077c44

acicovic committed:
Recommendations Block: Update @since for register_block()

acicovic committed:
Merge branch 'try/use-block-json-for-block-registration' of github.com:Parsely/wp-parsely into try/use-block-json-for-block-registration
